node/src/components/deploy_acceptor.rs

```diff
@@ -272,10 +273,34 @@ impl DeployAcceptor {
             }
         }
 
+        // If this has been received from the speculative exec server, use the block specified in
+        // the request, otherwise use the highest complete block.
```

Comment:
In the **non-**speculative flow, if the request is missing we use the `get_highest_complete_block_header_from_storage()`, which uses the `Storage::completed_blocks` to find the most recent completed block.

In the speculative flow, when we don't provide a block, we'll use the `get_highest_block_with_metadata_from_storage()`, which uses the `Storage::block_height_index` to get the most recent block.

I was wondering if the latter should also use the most recent **complete** block?

Reply:
Good catch!

Fixed in [3eb3d54](https://github.com/casper-network/casper-node/pull/3972/commits/3eb3d54ad72c10fdc7f47479d78683966d95a2fc).